### PR TITLE
don't use inplace backward, remove unnecessary zero for grad_input

### PR DESCRIPTION
**Pull request by ngimel**
Don't use inplace backward to avoid cloning, remove .zero_() for out-of-place.

### CODE REVIEW

File: torch/nn/_functions/thnn/auto.py [hunk 1]
```diff
@@ -164,10 +164,18 @@ def backward(self, grad_output):
                 additional_args = (output,) + additional_args
 
             if is_inplace and self.inplace:
+                inplace_tmp = False 
+                if additional_args[-1]==True:
+                     tmp_args = list(additional_args)
+                     tmp_args[-1] = False
+                     additional_args = tuple(tmp_args)
+            else:
+                inplace_tmp = False
```
Comment (apaszke) [inline torch/nn/_functions/thnn/auto.py]: is `inplace_tmp` ever going to be true?

File: torch/nn/_functions/thnn/auto.py [hunk 2]
```diff
@@ -164,10 +164,18 @@ def backward(self, grad_output):
                 additional_args = (output,) + additional_args
 
             if is_inplace and self.inplace:
+                inplace_tmp = False 
+                if additional_args[-1]==True:
```
Comment (apaszke) [inline torch/nn/_functions/thnn/auto.py]: linter will complain about lack of whitespace between around the operator

Comment (apaszke) [inline torch/nn/_functions/thnn/auto.py]: btw is it ever going to be `False` if we know that `self.inplace` is True?

File: torch/nn/_functions/thnn/auto.py (hunk 1 above)
Comment (ngimel) [inline torch/nn/_functions/thnn/auto.py]: No, it's not. I guess I'll remove that branch.

File: torch/nn/_functions/thnn/auto.py (hunk 2 above)
Comment (ngimel) [inline torch/nn/_functions/thnn/auto.py]: Idk. It shouldn't, but I haven't really checked the arguments for all backend functions.

**Comment by ngimel** [comment]
This should be good now.

**Comment by apaszke** [comment]
The tests are passing, but the linter complains 😕

**Review by apaszke — APPROVED** [review]

**Pull request merged**